### PR TITLE
[sw/silicon_creator] Shutdown cleanups

### LINKED ISSUE #7148: [mask_rom] Shutdown cleanups
The following cleanups need to be applied to the shutdown code before it can be considered complete:

- [x] Choose OTP-acceptable values (ie: constants with high hamming distances) for `shutdown_error_redact_t`, `alert_class_t`, `alert_enable_t`,  and `alert_escalate_t`. Closed by #8330
- [x] Implement FLASH_DISABLE.
- [x] Implement final ePMP restrictions for shutdown. (#14300)
- [x] Bring "local alert" configuration back into shutdown_init.
- [x] Refactor mocks in `shutdown_unittest.h` into their own mock headers.

### PR DESCRIPTION
**Pull request by alphan**
* **[sw/silicon_creator] Reset before disabling flash for proper shutdown in ROM_EXT:** Currently, shutdowns in ROM_EXT result in a hang instead of resetting the chip since flash_ctrl is disabled before we request a reset. This change adds a reset request before disabling flash_ctrl so that shutdowns reset the chip in ROM_EXT as well.

* **[sw/silicon_creator] Restrict ePMP in shutdown_hang():** This commit updates the ePMP configuration `shutdown_hang()` so that the only RX region is the inline asm block in that function. TOR configuration uses labels so that the same library can be used in ROM_EXT as well.

Disassemblies: [ROM](https://gist.github.com/alphan/5a24482469452d316d3524b392dde206), [ROM_EXT](https://gist.github.com/alphan/e105d00ba4a5e6e110cc4cb5ab94300b).

Fixes #7148

In ROM (RX enabled only for [0xcdb4, 0xce14)):
```
    cda4:                       80000737                lui     a4,0x80000
    cda8:                       177d                    addi    a4,a4,-1
    cdaa:                       009897b7                lui     a5,0x989
    cdae:                       d0078793                addi    a5,a5,-768 # 988d00 <_rom_ext_virtual_size+0x908d00>
    cdb2:                       4509                    li      a0,2
shutdown_hang():
/proc/self/cwd/sw/device/silicon_creator/lib/shutdown.c:447
  asm volatile(
    cdb4:                   /-> cacc                    sw      a1,20(a3)
    cdb6:                   |   00c82223                sw      a2,4(a6)
    cdba:                   |   00000297                auipc   t0,0x0
    cdbe:                   |   ffa28293                addi    t0,t0,-6 # cdb4 <shutdown_finalize+0x238>
    cdc2:                   |   0022d293                srli    t0,t0,0x2
    cdc6:                   |   3b029073                csrw    pmpaddr0,t0
    cdca:                   |   00000297                auipc   t0,0x0
    cdce:                   |   04a28293                addi    t0,t0,74 # ce14 <_text_end>
    cdd2:                   |   0022d293                srli    t0,t0,0x2
    cdd6:                   |   3b129073                csrw    pmpaddr1,t0
    cdda:                   |   3b271073                csrw    pmpaddr2,a4
    cdde:                   |   3a079073                csrw    pmpcfg0,a5
    cde2:                   |   3a101073                csrw    pmpcfg1,zero
    cde6:                   |   3a201073                csrw    pmpcfg2,zero
    cdea:                   |   3a301073                csrw    pmpcfg3,zero
    cdee:                   |   74751073                csrw    0x747,a0
    cdf2:                   |   10500073                wfi
    cdf6:                   |   10500073                wfi
    cdfa:                   |   10500073                wfi
    cdfe:                   |   10500073                wfi
    ce02:                   +-- bf4d                    j       cdb4 <shutdown_finalize+0x238>
    ce04:                   |   10500073                wfi
    ce08:                   |   10500073                wfi
    ce0c:                   +-- b765                    j       cdb4 <shutdown_finalize+0x238>
    ce0e:                   |   10500073                wfi
    ce12:                   \-- b74d                    j       cdb4 <shutdown_finalize+0x238>
```

In ROM_EXT (RX enabled only for [0x20002358, 0x200023b8)):
```
20002348:                       80000737                lui     a4,0x80000
2000234c:                       177d                    addi    a4,a4,-1
2000234e:                       009897b7                lui     a5,0x989
20002352:                       d0078793                addi    a5,a5,-768 # 988d00 <_owner_virtual_size+0x908d00>
20002356:                       4509                    li      a0,2
shutdown_hang():
/proc/self/cwd/sw/device/silicon_creator/lib/shutdown.c:447
  asm volatile(
20002358:                   /-> cacc                    sw      a1,20(a3)
2000235a:                   |   00c82223                sw      a2,4(a6)
2000235e:                   |   00000297                auipc   t0,0x0
20002362:                   |   ffa28293                addi    t0,t0,-6 # 20002358 <shutdown_finalize+0x238>
20002366:                   |   0022d293                srli    t0,t0,0x2
2000236a:                   |   3b029073                csrw    pmpaddr0,t0
2000236e:                   |   00000297                auipc   t0,0x0
20002372:                   |   04a28293                addi    t0,t0,74 # 200023b8 <_text_end>
20002376:                   |   0022d293                srli    t0,t0,0x2
2000237a:                   |   3b129073                csrw    pmpaddr1,t0
2000237e:                   |   3b271073                csrw    pmpaddr2,a4
20002382:                   |   3a079073                csrw    pmpcfg0,a5
20002386:                   |   3a101073                csrw    pmpcfg1,zero
2000238a:                   |   3a201073                csrw    pmpcfg2,zero
2000238e:                   |   3a301073                csrw    pmpcfg3,zero
20002392:                   |   74751073                csrw    0x747,a0
20002396:                   |   10500073                wfi
2000239a:                   |   10500073                wfi
2000239e:                   |   10500073                wfi
200023a2:                   |   10500073                wfi
200023a6:                   +-- bf4d                    j       20002358 <shutdown_finalize+0x238>
200023a8:                   |   10500073                wfi
200023ac:                   |   10500073                wfi
200023b0:                   +-- b765                    j       20002358 <shutdown_finalize+0x238>
200023b2:                   |   10500073                wfi
200023b6:                   \-- b74d                    j       20002358 <shutdown_finalize+0x238>
```

Signed-off-by: Alphan Ulusoy <alphan@google.com>

### CODE REVIEW

**Comment by alphan** [comment]
@mundaym can you review the inline asm that configures the ePMP? Thanks!

File: sw/device/silicon_creator/lib/shutdown.c [hunk 1]
```diff
@@ -445,30 +446,46 @@ SHUTDOWN_FUNC(noreturn, shutdown_hang(void)) {
...
-        // ROM may execute.  mundaym's suggestion: set entry 2 as a NAPOT
-        // region covering the entire address space, clear all its permission
-        // bits and set the lock bit, and then finally disable RLB to prevent
-        // any further modifications.
+        // Reconfigure ePMP such that only this inline asm is executable:
+        // - Entry 0: TOR, LRX, [.L_shutdown_hang_asm_start,
+        // .L_shutdown_hang_asm_end).
+        // - Entry 1: NAPOT, L, entire address space.
+        // - MSECCFG: MMWP set RLB unset to prevent any further modifications.
+        "la   t0, .L_shutdown_hang_asm_start;"
```
Comment (mundaym) [inline sw/device/silicon_creator/lib/shutdown.c:460]: IIRC the address needs to be shifted right by two (i.e. word aligned) before being loaded into the pmpaddrN registers.

File: sw/device/silicon_creator/lib/shutdown.c [hunk 2]
```diff
@@ -445,30 +446,46 @@ SHUTDOWN_FUNC(noreturn, shutdown_hang(void)) {
...
         :
         : [kSramRenewKey] "r"(1 << SRAM_CTRL_CTRL_RENEW_SCR_KEY_BIT),
           [kSramCtrlCtrlReg] "I"(SRAM_CTRL_CTRL_REG_OFFSET),
           [kSramCtrlBase] "r"(kSramCtrlBase),
           [kMultiBitBool4True] "r"(kMultiBitBool4True),
+          [kRstmgrResetReqReg] "I"(RSTMGR_RESET_REQ_REG_OFFSET),
           [kRstmgrBase] "r"(kRstmgrBase),
-          [kRstmgrResetReqReg] "I"(RSTMGR_RESET_REQ_REG_OFFSET));
+          [kNapotEntireAddressSpace] "r"(0x7fffffff),
+          [kPmpCfg0] "r"((EPMP_CFG_A_TOR | EPMP_CFG_LRX) |
```
Comment (mundaym) [inline sw/device/silicon_creator/lib/shutdown.c:485]: If you are using pmpaddr0 and pmpaddr1 for a TOR region then you want the config in pmp1cfg (i.e. shifted left by 8). Right now this is configuring pmp0cfg which will use the implicit starting address 0 as the bottom-of-range and pmpaddr0 as the top-of-range.

File: sw/device/silicon_creator/lib/shutdown.c [hunk 3]
```diff
@@ -445,30 +446,46 @@ SHUTDOWN_FUNC(noreturn, shutdown_hang(void)) {
...
+                         (EPMP_CFG_A_NAPOT | EPMP_CFG_L) << 8),
```
Comment (mundaym) [inline sw/device/silicon_creator/lib/shutdown.c:486]: Similarly this should be in pmp2cfg (i.e. shifted right by 16).

File: sw/device/silicon_creator/lib/shutdown.c [hunk 4]
```diff
@@ -445,30 +446,46 @@ SHUTDOWN_FUNC(noreturn, shutdown_hang(void)) {
...
         // Request a system reset.
         "sw %[kMultiBitBool4True], %[kRstmgrResetReqReg](%[kRstmgrBase]);"
 
-        // TODO(lowRISC/opentitan#7148): restrict the ePMP such that only
-        // ROM may execute.  mundaym's suggestion: set entry 2 as a NAPOT
-        // region covering the entire address space, clear all its permission
-        // bits and set the lock bit, and then finally disable RLB to prevent
-        // any further modifications.
+        // Reconfigure ePMP such that only this inline asm is executable:
+        // - Entry 0: TOR, LRX, [.L_shutdown_hang_asm_start,
```
Comment (mundaym) [inline sw/device/silicon_creator/lib/shutdown.c:456]: s/0/1/ (see comment below)

File: sw/device/silicon_creator/lib/shutdown.c [hunk 5]
```diff
@@ -445,30 +446,46 @@ SHUTDOWN_FUNC(noreturn, shutdown_hang(void)) {
...
+        // .L_shutdown_hang_asm_end).
+        // - Entry 1: NAPOT, L, entire address space.
```
Comment (mundaym) [inline sw/device/silicon_creator/lib/shutdown.c:458]: s/1/2/

File: sw/device/silicon_creator/lib/shutdown.c (hunk 1 above)
Comment (alphan) [reply sw/device/silicon_creator/lib/shutdown.c:460]: you're right, these addresses are bits 33-2 of a 34-bit physical address. added `"srli t0, t0, 2;"` for both addresses

File: sw/device/silicon_creator/lib/shutdown.c (hunk 2 above)
Comment (alphan) [reply sw/device/silicon_creator/lib/shutdown.c:485]: Done, thanks!

**Comment by alphan** [comment]
Thanks for the quick review @mundaym! I addressed your comments + removed to outer loop to remove instructions beyond the restricted range. PTAL

**Review by cfrantz — APPROVED** [review]

**Pull request merged**